### PR TITLE
 Add an opt-in to store incoming edges in `VecGraph` + misc

### PR DESCRIPTION
**Pull request by WaffleLapkin**
r? @fmease 

needed for #123939

### CODE REVIEW

File: compiler/rustc_data_structures/src/graph/vec_graph/mod.rs [hunk 1]
```diff
@@ -1,99 +1,235 @@
...
+    //
+    // `node_starts[n]` is the start of the list of targets of outgoing edges for node `n`.
+    // So you can get node's successors with `edge_targets[node_starts[n]..node_starts[n + 1]]`.
+    //
+    // If `BR` is true (back references are enabled), then `node_starts[n + edge_count]` is the
+    // start of the list of *sources* of incoming edges. You can get predecessors of a node
+    // similarly to its successors but offsetting by `edge_count`. `edge_count` is
+    // `edge_targets.len()/2` (again, in case BR is true) because half of the vec is back refs.
+    //
+    // All of this might be confusing, so here is an example graph and its representation:
```
Comment (wesleywiser) [inline compiler/rustc_data_structures/src/graph/vec_graph/mod.rs:24]: Nice diagram! 💯

File: compiler/rustc_data_structures/src/graph/vec_graph/mod.rs [hunk 2]
```diff
@@ -1,99 +1,235 @@
-use crate::graph::{DirectedGraph, NumEdges, Successors};
+use crate::graph::{DirectedGraph, NumEdges, Predecessors, Successors};
 use rustc_index::{Idx, IndexVec};
 
 #[cfg(test)]
 mod tests;
```
Comment (wesleywiser) [inline compiler/rustc_data_structures/src/graph/vec_graph/mod.rs:5]: nit: could you add a basic test or two for the predecessors functionality? (I think you also need to update some of the existing tests to compile with the extra generic parameter) 🙂

**Review by wesleywiser — APPROVED** [review]

File: compiler/rustc_data_structures/src/graph/vec_graph/mod.rs (hunk 1 above)
Comment (WaffleLapkin) [reply compiler/rustc_data_structures/src/graph/vec_graph/mod.rs:24]: Thanks!

**Pull request merged**